### PR TITLE
Non-strict mode for IDs

### LINKED ISSUE #47: message::parse_many_ids() fails to parse ID lists without spaces
Hello.
I have run into an issue: `message::parse_many_ids()` can't parse ID strings which don't contain a space separator, like this one:
`<029601d56c86$60e6e5e0$22b4b1a0$@mail.ru><02d901d56c96$89591ab0$9c0b5010$@mail.ru><114a0364-0ff7-f9ca-942a-3f58e45be343@otherdomain.tld>`

It's a great library which implements almost all features I need, but right now it's not possible for me to fetch some e-mails due to an exception thrown: `Parsing failure of the message ID`. Thunderbird fetches such messages successfully.

Please, is there anything you could do to support such space-lacking ID lists as this seems to be a common misbehaviour of mail servers?

I've read another issue (#7) and saw you mentioned 'strict' mode which can be turned off to lift some restrictions. I did so and it didn't resolve the issue. I guess, this doesn't apply to parsing ID lists. Could you please extend the non-strict mode to ID lists?
Thank you.

### LINKED ISSUE #50: Empty values of 'References' and 'In-Reply-To' headers cause exceptions
Hello.
It seems that `References` and `In-Reply-To` headers are treated the same way as `Message-id` is: if the header has empty value or a value that can't be parsed as a correct one, an exception is thrown which prevents emails with such headers from getting fetched by the library. But using an empty string in these headers seems to be a common behaviour among mail clients/servers. I have few emails sent with MS Outlook which have empty `References`, `In-Reply-To` and even `Message-id` headers, e.g. (emphasis is mine):
>...
>References: 
>In-Reply-To: 
>Subject: =?koi8-r?B?UkU6IOvBzcXS2Q==?=
>Date: Mon, 9 Dec 2019 14:14:31 +0300
>Message-ID: <004101d5ae81$d43895b0$7ca9c110$@somedomain.tld>
>MIME-Version: 1.0
>Content-Type: multipart/alternative;
>	boundary="----=_NextPart_000_0042_01D5AE9A.F9895020"
>**X-Mailer: Microsoft Outlook 15.0**
>...

Is this somehow possible to make parsing less restrictive and allow empty IDs in corresponding headers?
I saw that the code relies on throwing an exception and then tries to access first element from a vector of IDs returned by the `mime::parse_many_ids()` method.

It probably would be better not to throw exception in case when ID is empty, and then check if the returned vector of IDs parsed by `mime::parse_many_ids()` is not empty. If it is, just consider using an empty ID later on. I could implement it in a PR if you're OK with this. Otherwise, messages sent by MS Outlook and some other clients can't be accessed by the library.

### PR DESCRIPTION
**Pull request by tsilia**
Fixes #47
Fixes #50

### CODE REVIEW

File: src/message.cpp [hunk 1]
```diff
@@ -614,7 +619,12 @@ void message::parse_header_line(const string& header_line)
     else if (iequals(header_name, MESSAGE_ID_HEADER))
     {
         auto ids = parse_many_ids(header_value);
-        _message_id = ids[0];
+		if (_strict_mode && ids.empty())
```
Comment (karastojko) [inline src/message.cpp:622]: The strict mode has to be negated?

File: src/message.cpp [hunk 2]
```diff
@@ -1318,9 +1328,28 @@ string message::format_many_ids(const string& id)
...
+		auto start = ids.cbegin();
+		auto end = ids.cend();
+		match_flag_type flags = match_default;
+		match_results<string::const_iterator> what;
+		while (regex_search(start, end, what, r, flags))
+		{
+			idv.push_back(what[0]);
+			start = what[0].second;
+			flags |= match_prev_avail;
+			flags |= match_not_bob;
```
Comment (karastojko) [inline src/message.cpp:1348]: Is it necessary to set the flags in each loop iteration? In addition, are these flags necessary at all or the default one plus the not null flag is good enough?

File: src/message.cpp [hunk 3]
```diff
@@ -1318,9 +1328,28 @@ string message::format_many_ids(const string& id)
...
+	}
+	else if (!_strict_mode)
+	{
+		auto start = ids.cbegin();
+		auto end = ids.cend();
+		match_flag_type flags = match_default;
+		match_results<string::const_iterator> what;
+		while (regex_search(start, end, what, r, flags))
+		{
+			idv.push_back(what[0]);
```
Comment (karastojko) [inline src/message.cpp:1345]: The angle brackets are not removed from the ID as in the loop below. The idea for the ID is not to keep the brackets once it is parsed.

File: src/message.cpp [hunk 4]
```diff
@@ -1318,9 +1328,28 @@ string message::format_many_ids(const string& id)
...
+			start = what[0].second;
+			flags |= match_prev_avail;
+			flags |= match_not_bob;
+		}
+		return idv;
+	}
     split(idv, ids, is_any_of(codec::SPACE_STR), boost::algorithm::token_compress_on);
-    const regex r(R"(<([a-zA-Z0-9\!#\$%&'\*\+\-\./=\?\^\_`\{\|\}\~]+)\@([a-zA-Z0-9\!#\$%&'\*\+\-\./=\?\^\_`\{\|\}\~]+)>)");
     smatch m;
     for (auto& id : idv)
```
Comment (karastojko) [inline src/message.cpp:1354]: With the loop on `regex_search` I believe this loop can be removed.

**Comment by karastojko** [comment]
@tsilia 
I have merged the two loops into one, please review the [PR55](https://github.com/karastojko/mailio/pull/55).

**Comment by tsilia** [comment]
@karastojko Good points. Sorry for messing up the PR, I focused on getting ID lists without spaces parsed and totally forgot to complete and test the code. We probably should just close this and go on with your #55. Thanks for your time.
There is a concern though I'll post about there.

**Pull request merged**